design/authorization-handler.md

```diff
@@ -0,0 +1,590 @@
+
+In the next text, it can be assumed that Authorization Handler and Client Scope are the same entity for all purposes. So by Client Scope is meant the "Dynamic" client scope, which will be able to handle dynamic scopes and RAR requests. Not just "Static client scope" as available in Keycloak in version 15.0.2.
+
+**Note**: We need community feedback on what the best path would be, and if we decide to go with the authorization handler, it would also be great to get a better name for it.
+
+## The Authorization Handler
+
+A new type of entity will be created in Keycloak that will allow users to configure the behaviour of Static, Dynamic Scopes and RAR in a composable way.
+
+The AuthorizationHandler will be a composite structure consisting of the following pluggable components:
+
+* The AuthorizationContext initiator
+* A Compliance processor
+* A consent screen enhancer / builder
+* Configured Protocol Mappers.
+
+
+### AuthorizationContext initiator
+
+At the initial step, the authorization data sent by the client can be retrieved from the initial OIDC Authorization-Endpoint request (or Token-Request in the use-cases like CIBA or OAuth2 resource-owner-password-credentials grant). This retrieved data will be referred to as AuthorizationRequest, which means the initial request to Keycloak to start the authentication process.
+
+**NOTE: This design consider just OpenID Connect specification for now (Dynamic scopes and RAR will be OIDC specific when SAML would still only the support for default client scopes as it has today in Keycloak 15 or earlier)**
+
+### How to parse authorization data
+
+The authorization data is available in the OIDC request by 2 ways:
+
+* In the scope parameter. The scope parameter is already supported in current Keycloak, but rather in static way
+
+* In the authorization_details parameter as defined in the [RAR specification](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-rar-07).
+
+Once Keycloak parses the data from any of these 2 parameters, it will save them in the AuthorizationContext object. The idea is to unify the data in the AuthorizationContext in some structure, which will serve as an abstraction layer over the data that was sent via the scope parameter or authorization_details. However Keycloak will still need to have some flag in AuthorizationContext to track how the data were sent. This may be needed for various reasons. 
+
+For example [Grant Management API specification - section 6.4](https://openid.net/specs/fapi-grant-management-01.html#section-6.4) (not yet supported by Keycloak) allows clients to query Keycloak about the granted consents. Note that the returned structure contains both scope and authorization_details and hence it is needed by KC to track how the particular request was sent.
+
+### AuthorizationContext object
```

Comment:
I'm afraid that the current keycloak's authorization service has already use the `AuthorizationContext` class.

IMO, it might be better to use different class name (EG. `OAuth2AuthorizationContext`)

Reply:
@mposolda @tnorimat Apparently `ClientAuthorizationContext` is already being used by the `Authorization Services` module. We'll need to come up with a different name for this context.

Maybe `RequestAuthorizationContext`, `IdentityAuthorizaitonContext`?

Reply:
@dgozalo `RequestAuthorizationContext` looks good to me as it is clear from it, that it is request-scoped object. One more idea is `ClientRequestContext` (does not exists in Keycloak, but in Resteasy dependency). Leaving to you what you prefer...

Reply:
Actually, there's a class called `AuthorizationRequestContext` used in https://github.com/keycloak/keycloak/blob/31345c49b12352639de9406ccefd51f19afdf426/services/src/main/java/org/keycloak/protocol/oidc/endpoints/AuthorizationEndpoint.java#L167.

It's used in the OIDC protocol endpoint but it does something quite similar to what we have in mind, even thought this one will be more generic, to be used in other grants and specifications. 

Maybe we could "generify" and reuse this class for our purposes instead of creating a new one.

Reply:
I am not 100% sure as `AuthorizationRequestContext` is currently used for client policies and it is part of some infrastructure for client policies. Re-use the same class for both client policies and this new functionality might be confusing and doesn't help with very clear "separation of responsibilities" in the sources IMO.

Reply:
That's true. I was't sure what it was being used for, but it's clear that it's doing a completely different thing and we shouldn't reuse.